index/segment/mem/segment_simple.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// New returns a new in-memory index.
+func New(id segment.ID, opts Options) (Segment, error) {
+	seg := &simpleSegment{
+		opts:      opts,
+		id:        id,
+		docIDGen:  atomic.NewUint32(0),
+		termsDict: newSimpleTermsDictionary(opts),
+	}
+	seg.docs.values = make([]document, opts.InitialCapacity())
+	searcher := newSequentialSearcher(seg, differenceNegationFn, opts.PostingsListPool())
+	seg.searcher = searcher
+	return seg, nil
+}
+
+// TODO(prateek): consider copy semantics for input data, esp if we can store it on the native heap
+func (i *simpleSegment) Insert(d doc.Document) error {
+	newDoc := document{
+		Document: d,
+		docID:    segment.DocID(i.docIDGen.Inc()),
+	}
+	i.insertDocument(newDoc)
+	return i.insertTerms(newDoc)
+}
+
+func (i *simpleSegment) insertDocument(doc document) {
+	docID := int64(doc.docID)
+	// can early terminate if we have sufficient capacity
+	i.docs.RLock()
+	size := len(i.docs.values)
+	if int64(size) > docID {
+		// NB(prateek): only need a Read-lock here despite an insert operation because
+		// we're guranteed to never have conflicts with docID (it's monotonoic increasing),
+		// and have checked `i.docs.values` is large enough.
+		i.docs.values[doc.docID] = doc
+		i.docs.RUnlock()
+		return
+	}
+	i.docs.RUnlock()
+
+	// need to expand capacity
+	i.docs.Lock()
+	size = len(i.docs.values)
+	// expanded since we released the lock
+	if int64(size) > docID {
+		i.docs.values[doc.docID] = doc
+		i.docs.Unlock()
+		return
+	}
+
+	docs := make([]document, 2*(size+1))
+	copy(docs, i.docs.values)
+	i.docs.values = docs
+	i.docs.values[doc.docID] = doc
+	i.docs.Unlock()
+}
+
+func (i *simpleSegment) insertTerms(doc document) error {
+	// insert each of the indexed fields into the reverse index
+	// TODO: current implementation allows for partial indexing. Evaluate perf impact of not doing that.
+	for _, field := range doc.Fields {
+		if err := i.termsDict.Insert(field, doc.docID); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (i *simpleSegment) Query(query segment.Query) (segment.ResultsIter, error) {
+	ids, pendingFn, err := i.searcher.Query(query)
+	if err != nil {
+		return nil, err
+	}
+
+	return newResultsIter(ids, pendingFn, i), nil
+}
+
+func (i *simpleSegment) Filter(
+	f segment.Filter,
+) (segment.PostingsList, matchPredicate, error) {
+	docs, err := i.termsDict.Fetch(f.FieldName, f.FieldValueFilter, termFetchOptions{isRegexp: f.Regexp})
+	return docs, nil, err
+}
+
+func (i *simpleSegment) Delete(d doc.Document) error {
+	panic("not implemented")
+}
+
+func (i *simpleSegment) Size() uint32 {
+	return i.docIDGen.Load()
+}
+
+func (i *simpleSegment) ID() segment.ID {
+	return i.id
+}
+
+func (i *simpleSegment) Options() Options {
+	return i.opts
+}
+
+func (i *simpleSegment) FetchDocument(id segment.DocID) (document, error) {
+	if uint32(id) > i.Size() {
+		return document{}, errUnknownDocID
+	}
+
+	i.docs.RLock()
+	d := i.docs.values[id]
+	i.docs.RUnlock()
+	return d, nil
```

Comment:
Hm so I believe this method is a little racey, the `docIDGen` is incremented before the slice is grown potentially.  Wouldn't it just be safer to grab the RLock() and check the length of the slice?  You have to grab the RLock() in the case that its valid anyway.

Reply:
Yeah, good catch. Thanks!

Reply:
done